### PR TITLE
#187300145 User Logout 

### PR DESCRIPTION
**Pull request by Sophie20202**
# What does this PR do?
This PR implements the functionality for users to Logout

# Description of Task to be completed?
A user should be able to logout of the application

# How should this be manually tested?
npm run install
npm run dev
then swagger for testing

# Any background context you want to provide?
This servise uses Bearer token  as an authorization in header to allow user logout

# What are the relevant pivotal tracker/Trello stories?
-Story ID: #187300145

### CODE REVIEW

**Review by musabehonore — APPROVED** [review]

**Review by fab-ryan — CHANGES_REQUESTED** [review]
Please squash those commits

File: src/controllers/logoutController.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { Request, Response } from 'express';
+import { sendResponse } from '../utils';
+import Database from '../database';
+
+const logoutUser = async (req: Request, res: Response) => {
+  try {
+    const authorization = req.header('Authorization')?.split(' ')[1];
+    if (!authorization) {
+      return sendResponse(res, 403, ' ', 'please login ...');
+    }
```
Comment (fab-ryan) [inline src/controllers/logoutController.ts]: remove this into middlware

**Review by fab-ryan — CHANGES_REQUESTED** [review]
- no middleware that checks if there token is valid or blacklisted
- return result was not what it's supposed too
- every endpoint that have auth must have that endpoint

File: src/middleware/authentication.ts [hunk 2]
```diff
@@ -11,10 +12,18 @@ export const isAuthenticated = (
...
         return sendResponse(res, 403, null, 'You are not authorized');
       }
+      const authorization = req.header('Authorization')?.split(' ')[1];
+
+      const logout = await Database.Logout.findOne({
+        where: { token: authorization },
+      });
+      if (logout) {
+        return sendResponse(res, 403, null, 'not authorized');
+      }
```
Comment (fab-ryan) [inline src/middleware/authentication.ts]: and again they told you to use function not implement it in here

**Review by fab-ryan — CHANGES_REQUESTED** [review]
please do those test well because lines are not tested and what @denislohan suggested you didn't implement it

File: src/middleware/authentication.ts (hunk 2 above)
Comment (Sophie20202) [reply src/middleware/authentication.ts]: I understand when i use a function i will implement it where?

File: src/middleware/authentication.ts [hunk 3]
```diff
@@ -11,10 +12,16 @@ export const isAuthenticated = (
...
+    async (err: Error, user: UserAttributes) => {
       if (!user) {
         return sendResponse(res, 403, null, 'You are not authorized');
       }
+      const token = req.header('Authorization')?.split(' ')[1];
+      const logout = await Database.Logout.findOne({ where: { token } });
+      if (logout) {
+        return sendResponse(res, 401, null, 'Token already blacklisted');
+      }
+
```
Comment (fab-ryan) [inline src/middleware/authentication.ts]: please separaete codes

**Review by fab-ryan — CHANGES_REQUESTED** [review]
please separate your implementation

File: src/middleware/authentication.ts (hunk 3 above)
Comment (fab-ryan) [inline src/middleware/authentication.ts]: please this is not implementation we need

**Review by fab-ryan — CHANGES_REQUESTED** [review]
none of the implementation requested was done?

**Review by Nadinefiona — APPROVED** [review]

**Review by fab-ryan — APPROVED** [review]

**Pull request merged**